### ISSUE TITLE
DNSQuestion's udpAnswer length is not equal to UDP response packet payload length anymore

### ISSUE THREAD

**Issue by dshcherb** [PowerDNS/pdns#11771]
- Program: Recursor
 - Issue type: Bug report

### Short description

After upgrading to 4.7.0 I found that some of my Lua scripts that initiate a [UDP call](https://doc.powerdns.com/recursor/lua-scripting/hooks.html#udp-query-response) to an external service stopped working correctly (the external service itself has not changed). Namely, the udpAnswer string comparison checks started to fail. Digging more showed that the length of udpAnswer is not equal to the UDP response payload size anymore and there is garbage at the end of udpAnswer after the UDP payload.

See "Other information" for a detailed overview.

### Environment
 - Operating system: OpenWRT 22.03.0-rc5
 - Software version: 4.7.0
 - Software source: Operating system repository

### Steps to reproduce
1. Set up a `recursor.conf` such that it contains a custom DNS script `lua-dns-script=/etc/powerdns/dns-script.lua`;
2. Make a query to an external UDP service following https://doc.powerdns.com/recursor/lua-scripting/hooks.html#udp-query-response
3. Observe that the [udpAnswer](https://doc.powerdns.com/recursor/lua-scripting/dq.html#DNSQuestion.udpAnswer) field has a length larger than the UDP packet payload length.

### Expected behaviour

Previously `udpAnswer` length was equal to the length of the response UDP packet payload.

### Actual behaviour

udpAnswer length is 512 and there is garbage at the end of the string causing predicates based on string contents to fail.

### Other information

After upgrading to 4.7.0 I found that some of my Lua scripts that initiate a [UDP call](https://doc.powerdns.com/recursor/lua-scripting/hooks.html#udp-query-response) to an external service stopped working correctly (the external service itself has not changed).

I found that instead of simply getting a string response like "ok" which my service is sending I started seeing the following in the recursor log:

```
# the log line in the script: pdnslog("external service responded with: "..dq.udpAnswer, pdns.loglevels.Error)
# outputs:
external service responded with: ok
external service responded with: ok^\
external service responded with: ok^X<B6>^A
external service responded with: ok^E#^E<A0>^^e<E9>ʞ<F5>ۈ^\:;<A5><B4>ET<BF>^B8<97>h<DD><E4>־<C4>F9<8C>
<DB><D9>
<AB><83>:<EB>^K<BF>
# ... and so on with more garbage
```

Doing a tcpdump shows me that an external service listening on 8054 returns 3 bytes: `ok\0`. I added null byte termination in the external service just to check if this will help determining the boundaries (it's not a stream anyway so I should not need it) but Lua doesn't treat them as string terminators https://www.lua.org/manual/5.1/manual.html#2.1 ("embedded zeros").

```
tcpdump -X -i lo port 8054

23:21:10.730359 IP localhost.8054 > localhost.52302: UDP, length 3
	0x0000:  4500 001f 0016 4000 4011 3cb6 7f00 0001  E.....@.@.<.....
	0x0010:  7f00 0001 1f76 cc4e 000b fe1e 6f6b 00    .....v.N....ok.
```

Printing the length of the [udpAnswer](https://doc.powerdns.com/recursor/lua-scripting/dq.html#DNSQuestion.udpAnswer) shows 512 instead of 3:

```
# pdnslog("answer len: "..tostring(string.len(dq.udpAnswer)), pdns.loglevels.Error)

Jul 10 23:21:13 myhost pdns_recursor[10356]: answer len: 512
```

Potentially something changed in https://github.com/PowerDNS/pdns/commit/2a17f6c6c25fb3bc6eabf4542421850652ea92b9 which now results in this behavior.

**Comment by omoerbeek** [comment]
Thanks for the report! Which version where you running before?

**Comment by omoerbeek** [comment]
This is very likely the fix. I will write a regression test for this and then create a PR.
```
diff --git a/pdns/pdns_recursor.cc b/pdns/pdns_recursor.cc
index ad3b26304..1f9d62aad 100644
--- a/pdns/pdns_recursor.cc
+++ b/pdns/pdns_recursor.cc
@@ -215,6 +215,7 @@ static void handleGenUDPQueryResponse(int fd, FDMultiplexer::funcparam_t& var)
 
   t_fdm->removeReadFD(fd);
   if (ret >= 0) {
+    resp.resize(ret);
     MT->sendEvent(pident, &resp);
   }
   else {
```

**Comment by omoerbeek** [comment]
Fix is merged to master, backports wil follow soon.

**Comment by dshcherb** [comment]
Thanks for the blazingly fast reply and fix! :^)

> Thanks for the report! Which version where you running before?

It was 4.4.2 included in the 21.02 release of OpenWRT.
